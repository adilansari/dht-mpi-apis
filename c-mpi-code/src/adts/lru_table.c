
#include "lru_table.h"

static int
max(int i1, int i2)
{
  return (i1>i2) ? i1 : i2;
}

/*
static int
min(int i1, int i2)
{
  return (i1<i2) ? i1 : i2;
}
*/

struct lru_table*
lru_table_create(int capacity, int limit)
{
  // SHOW_FII(capacity, limit);

  int i = 0;
  struct lru_table *new_table = NULL;

  new_table =
    (struct lru_table*) malloc(sizeof(struct lru_table));
  if (! new_table)
    return (NULL);

  new_table->size     = 0;
  new_table->capacity = capacity;
  new_table->limit    = limit;

  new_table->lru = dpkm_list_create(limit);

  new_table->array =
    (struct dpkm_list**) malloc(capacity * sizeof(struct dpkm_list*));
  if (!new_table->array)
  {
    free(new_table);
    return NULL;
  }

  for (i = 0; i < capacity; i++)
  {
    // These lists do not enforce a limit!
    struct dpkm_list* new_dpkm_list = dpkm_list_create(limit+3);
    if (! new_dpkm_list)
      return NULL;
    new_table->array[i] = new_dpkm_list;
  }

  return new_table;
}

/**
   Does not free the data.
*/
void
lru_table_free(struct lru_table* target)
{
  NOTE_F;
  int i;
  for (i = 0; i < target->capacity; i++)
    dpkm_list_free(target->array[i]);
  free(target->array);
  dpkm_list_free(target->lru);
  free(target);
}

/**
   Frees structure and the data.
*/
void
lru_table_destroy(struct lru_table* target)
{
  NOTE_F;
  int i;
  for (i = 0; i < target->capacity; i++)
    dpkm_list_destroy(target->array[i]);
  free(target->array);
  // Do not free the keys or data in this list- they are referred
  //    to by the array entries
  dpkm_list_free(target->lru);
  free(target);
}


/**
   Adds the key/data pair to the table.

   Adds new item to end of table list and lru list.
   Sets those two item ptrs to each other.

   @param key May not be NULL
   @param data

   @return NULL if the table is not overfull or a
           struct keyvalue* representing the key/data pair
           that was LRU and selected as a victim.
 */
struct keyvalue*
lru_table_add(struct lru_table* table, char* key,
              void* data, int length)
{
  NOTE_FS(key);
  assert(key);

  struct keyvalue* result = NULL;

  int index = hash_string(key, table->capacity);

  // printf("%s -> %i \n", key, index);

  dpkm_list_add(table->array[index], key, data, length);

  struct dpkm_list_item* victim =
    dpkm_list_add(table->lru, key, data, length);

  table->array[index]->tail->ptr = table->lru->tail;
  table->lru->tail->ptr          = table->array[index]->tail;

  if (victim == NULL)
  {
    table->size++;
  }
  else
  {
    int victim_index = hash_string(victim->key, table->capacity);
    dpkm_list_excise(table->array[victim_index], victim->ptr);
    result = (struct keyvalue*) malloc(sizeof(struct keyvalue));
    result->key    = victim->key;
    result->data   = victim->data;
    result->length = victim->length;
    free(victim->ptr);
    free(victim);
  }

  return result;
}

/**
   Search the table and touch the key/data pair
   to prevent vicitimization.

   @return Allocated keyvalue with pointers to key/data in table.
*/
struct keyvalue*
lru_table_search(struct lru_table* table, char* key)
{
  SHOW_FS(key);

  struct dpkm_list_item* item = NULL;
  int index = hash_string(key, table->capacity);
  struct keyvalue* result;

  item = dpkm_list_search(table->array[index], key);
  if (item)
  {
    dpkm_list_pushback(table->lru, item->ptr);
    result = malloc(sizeof(struct keyvalue));
    result->key    = key;
    result->data   = item->data;
    result->length = item->length;

    DEBUG(printdata("lru found: ", item->data, item->length));
    DEBUG(printf("\n"));

    return result;
  }

  return NULL;
}

/**
   Search the table for the key/data pair and update the data.
   Copies data out of value: does not free it.
   Frees old data pointer.
   @return True iff the item was found and updated.
*/
bool
lru_table_update(struct lru_table* table,
                 char* key, char* value, int offset, int length)
{
  SHOW_FSII(key,offset,length);

  struct dpkm_list_item* item = NULL;
  int index = hash_string(key, table->capacity);

  item = dpkm_list_search(table->array[index], key);
  if (item)
    dpkm_list_pushback(table->lru, item->ptr);
  else
    return false;

  gossip_do(MASK_ADTS, lru_table_printf("%s", table));
  DEBUG(printdata("old data: ", item->data, item->length));

  int new_length = max(item->length, offset+length);

  char* update = realloc(item->data, new_length);
  assert(update);
  memcpy(update+offset, value, length);

  item->data = update;
  item->length = new_length;
  item = item->ptr;
  item->data = update;
  item->length = new_length;

  DEBUG(printdata("new data: ", item->data, item->length));

  DEBUG(lru_table_printdata(table));

  return true;
}

/**
   Remove the oldest key/value pair.

   @return NULL if the table is empty or a
           struct keyvalue* representing the key/data pair
           that was LRU.
 */
struct keyvalue*
lru_table_poll(struct lru_table* table)
{
  // NOTE_F;

  struct keyvalue* result = NULL;

  struct dpkm_list_item* oldest = dpkm_list_poll(table->lru);

  if (oldest == NULL)
  {
    return NULL;
  }
  else
  {
    int oldest_index = hash_string(oldest->key, table->capacity);
    dpkm_list_excise(table->array[oldest_index], oldest->ptr);
    result = (struct keyvalue*) malloc(sizeof(struct keyvalue));
    // SHOW_S(oldest->key);
    result->key    = oldest->key;
    result->data   = oldest->data;
    result->length = oldest->length;
    free(oldest->ptr);
    free(oldest);
    table->size--;
  }

  return result;
}

/**
   Internal method to remove key from table, not lru list.
*/
/*
bool
lru_table_remove(struct lru_table* target, struct dpkm_list_item* ptr)
{

list_excise()
}
*/

/**
   Write the table to stdout.
   @param format specifies the output format for the data items
 */
void
lru_table_dump(char* format, struct lru_table* target)
{
  int i;
  printf("LRU_TABLE(%i) \n{\n", target->size);
  printf("LRU: ");
  dpkm_list_printf(format, target->lru);
  for (i = 0; i < target->capacity; i++)
    if (target->array[i]->size > 0)
    {
      printf(" %i: ", i);
      dpkm_list_printf(format, target->array[i]);
    }
  printf("}\n");
}

/**
   @param format specifies the output format for the data items
 */
void
lru_table_printf(char* format, struct lru_table* target)
{
  int i;
  printf("LRU_TABLE(%i) \n{\n", target->size);
  printf("LRU: ");
  dpkm_list_printf(format, target->lru);
  for (i = 0; i < target->capacity; i++)
    if (target->array[i]->size > 0)
    {
      printf(" %i: ", i);
      dpkm_list_printf(format, target->array[i]);
    }
  printf("}\n");
}

/**
   @param format specifies the output format for the data items
 */
void
lru_table_fprintf(FILE* file, char* format, struct lru_table* target)
{
  int i;
  fprintf(file, "LRU_TABLE(%i) \n{\n", target->size);
  fprintf(file, "LRU: ");
  dpkm_list_fprintf(file, format, target->lru);
  for (i = 0; i < target->capacity; i++)
    if (target->array[i]->size > 0)
    {
      fprintf(file, " %i: ", i);
      dpkm_list_fprintf(file, format, target->array[i]);
    }
  fprintf(file, "}\n");
}

/*
void
lru_table_dumpkeys(struct lru_table* target)
{
  int i;
  printf("{\n");
  for (i = 0; i < target->capacity; i++)
    if (target->array[i]->size)
    {
      printf("%i:", i);
      dpkm_list_dumpkeys(target->array[i]);
    }
  printf("}\n");
}
*/

/**
 */
void
lru_table_fprintdata(FILE* file, struct lru_table* target)
{
  int i;
  fprintf(file, "LRU_TABLE(%i) \n{\n", target->size);
  fprintf(file, "LRU: ");
  dpkm_list_fprintdata(file, target->lru);
  for (i = 0; i < target->capacity; i++)
    if (target->array[i]->size > 0)
    {
      fprintf(file, " %i: ", i);
      dpkm_list_fprintdata(file, target->array[i]);
    }
  fprintf(file, "}\n");
}

void
lru_table_printdata(struct lru_table* target)
{
  lru_table_fprintdata(stdout, target);;
}


/** Dump dpkm_list to string a la snprintf()
    size must be greater than 2.
    format specifies the output format for the data items
    internally allocates O(size) memory
    returns int greater than size if size limits are exceeded
            indicating result is garbage
*/
int lru_table_tostring(char* str, size_t size,
                       char* format, struct lru_table* target)
{
  int   error = size+1;
  char* ptr   = str;
  int i;
  ptr += sprintf(str, "{\n");

  char* s = (char*) malloc(sizeof(char) * size);

  for (i = 0; i < target->capacity; i++)
  {
    int r = dpkm_list_tostring(s, size, format, target->array[i]);
    if ((ptr-str) + r + 2 < size)
      ptr += sprintf(ptr, "%s\n", s);
    else
      return error;
  }
  sprintf(ptr, "}\n");

  free(s);
  return (ptr-str);
}

/**
   Debugging only.
*/

#ifdef DEBUG_LRU_TABLE
int
main()
{
  struct lru_table* table = lru_table_create(11);

  lru_table_add(table, "key1", "val1");

  lru_table_dump("%s", table);

}
#endif
