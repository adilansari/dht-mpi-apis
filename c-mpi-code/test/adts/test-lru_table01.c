
/**
 * Test lru_table functionality.
 * */

#include <lru_table.h>

#include "test_helpers.h"

void extra_kv(struct keyvalue* kv)
{
  if (kv)
  {
    printf("extra: %s \n", kv->key);
    free(kv);
  }
}

/**
   Clean up extra keyvalue and its key and data.
*/
void extra_kv_kd(struct keyvalue* kv)
{
  if (kv)
  {
    printf("extra: %s \n", kv->key);
    if (kv->data)
      free(kv->data);
    if (kv->key)
      free(kv->key);
    free(kv);
  }
}

int
main(int argc, char* argv[])
{
  MPI_Init(&argc, &argv);
  whoami();

  struct lru_table* table = lru_table_create(5, 4);

  int* one   = malloc(sizeof(int));
  int* two   = malloc(sizeof(int));

  *one   = 1;
  *two   = 2;

  struct keyvalue* kv;

  kv = lru_table_add(table, "one", one, sizeof(int));
  lru_table_printf("%i", table);
  extra_kv(kv);

  printf("\n");

  kv = lru_table_add(table, "two", two, sizeof(int));
  lru_table_printf("%i", table);
  extra_kv(kv);

  printf("\n");

  kv = lru_table_search(table, "one");
  assert(kv);
  printf("data: %i\n", *(int*) kv->data);
  free(kv);

  kv = lru_table_search(table, "two");
  assert(kv);
  printf("data: %i\n", *(int*) kv->data);
  free(kv);

  kv = lru_table_search(table, "one");
  assert(kv);
  printf("data: %i\n", *(int*) kv->data);
  free(kv);

  while ((kv = lru_table_poll(table)))
  {
    printf("polled: %s %i\n", kv->key, *kv->data);
    free(kv);
  }

  assert(table->size == 0);
  lru_table_free(table);

  free(one);
  free(two);

  printf("loop...");
  table = lru_table_create(6, 4);
  for (int i = 0; i < 30; i++)
  {
    lru_table_printdata(table);
    char* key  = makekey(i);
    if (rand() % 2)
    {
      void* data = makedata(i);
      kv = lru_table_add(table, key, data, i);
      extra_kv_kd(kv);
    }
    else
    {
      printf("search: %s\n", key);
      kv = lru_table_search(table, key);
      free(key);
      extra_kv(kv);
    }
  }

  puts("draining...");
  while ((kv = lru_table_poll(table)))
  {
    printf("polled: %s\n", kv->key);
    printdata("", kv->data, kv->length);
    keyvalue_destroy(kv);
  }

  assert(table->size == 0);
  lru_table_free(table);

  /*
  int* result = (int*) lru_table_search(table, "five");
  printf("result: %i \n", *result);
  lru_table_dump("%i", table);

  result = (int*) lru_table_search(table, "twenty");
  if (result == NULL)
    printf("Got NULL \n");

  result = (int*) lru_table_search(table, "five");
  printf("result: %i \n", *result);
  lru_table_dump("%i", table);

  while ((kv = lru_table_poll(table)))
    printf("polled: %s \n", kv->key);
  */


  /*

  lru_table_add(table, "five", &five);
  lru_table_excise(table, table->head);
  lru_table_dump("%i", table);
  printf("size: %i \n", table->size);

  kv = lru_table_add(table, "eight", &eight);
  if (kv)
    printf("%s \n", kv->key);
  lru_table_dump("%i", table);
  printf("size: %i \n", table->size);

  lru_table_touch(table, table->head->next);
  lru_table_dump("%i", table);
  printf("size: %i \n", table->size);

  struct lru_table_item* item;
  item = lru_table_poll(table);
  lru_table_dump("%i", table);
  printf("size: %i \n", table->size);

  while ((item = lru_table_poll(table)))
    lru_table_dump("%i", table);

  */
  /*
  lru_table_dump("%i", table);
  lru_table_poll(table);
  lru_table_dump("%i", table);
  lru_table_pop(table);
  lru_table_dump("%i", table);
  lru_table_add(table, &seven);
  lru_table_add(table, &six);
  lru_table_dump("%i", table);
  */

  // lru_table_clobber(table);
  // lru_table_clear(table);
  // printf("size(table): %i \n", table->size);

  // lru_table_clear(K);
  // printf("size(K): %i \n", K->size);

  // lru_table_dump("%i", table);

  // dmalloc_shutdown();

  printf("\n");
  printf("Normal exit.\n");

  MPI_Finalize();

  return 0;
}
