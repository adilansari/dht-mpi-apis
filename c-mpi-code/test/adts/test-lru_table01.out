Debug to stdout
[0] lru_table_add(one)... 
LRU_TABLE(1) 
{
LRU: [(one,1)]
 1: [(one,1)]
}

[0] lru_table_add(two)... 
LRU_TABLE(2) 
{
LRU: [(one,1),(two,2)]
 1: [(one,1)]
 2: [(two,2)]
}

[0] lru_table_search(key=one)... 
lru found: [\1\0\0\0]
data: 1
[0] lru_table_search(key=two)... 
lru found: [\2\0\0\0]
data: 2
[0] lru_table_search(key=one)... 
lru found: [\1\0\0\0]
data: 1
polled: two 2
polled: one 1
[0] lru_table_free()... 
loop...LRU_TABLE(0) 
{
LRU: []
}
[0] lru_table_add(0)... 
LRU_TABLE(1) 
{
LRU: [(0@0,[])]
 4: [(0@0,[])]
}
search: 1
[0] lru_table_search(key=1)... 
LRU_TABLE(1) 
{
LRU: [(0@0,[])]
 4: [(0@0,[])]
}
[0] lru_table_add(2)... 
LRU_TABLE(2) 
{
LRU: [(0@0,[]),(2@2,[XX])]
 4: [(0@0,[])]
 5: [(2@2,[XX])]
}
[0] lru_table_add(3)... 
LRU_TABLE(3) 
{
LRU: [(0@0,[]),(2@2,[XX]),(3@3,[XXX])]
 4: [(0@0,[]),(3@3,[XXX])]
 5: [(2@2,[XX])]
}
[0] lru_table_add(4)... 
LRU_TABLE(4) 
{
LRU: [(0@0,[]),(2@2,[XX]),(3@3,[XXX]),(4@4,[XXXX])]
 0: [(4@4,[XXXX])]
 4: [(0@0,[]),(3@3,[XXX])]
 5: [(2@2,[XX])]
}
[0] lru_table_add(5)... 
extra: 0 
LRU_TABLE(4) 
{
LRU: [(2@2,[XX]),(3@3,[XXX]),(4@4,[XXXX]),(5@5,[XXXXX])]
 0: [(4@4,[XXXX])]
 4: [(3@3,[XXX]),(5@5,[XXXXX])]
 5: [(2@2,[XX])]
}
search: 6
[0] lru_table_search(key=6)... 
LRU_TABLE(4) 
{
LRU: [(2@2,[XX]),(3@3,[XXX]),(4@4,[XXXX]),(5@5,[XXXXX])]
 0: [(4@4,[XXXX])]
 4: [(3@3,[XXX]),(5@5,[XXXXX])]
 5: [(2@2,[XX])]
}
search: 7
[0] lru_table_search(key=7)... 
LRU_TABLE(4) 
{
LRU: [(2@2,[XX]),(3@3,[XXX]),(4@4,[XXXX]),(5@5,[XXXXX])]
 0: [(4@4,[XXXX])]
 4: [(3@3,[XXX]),(5@5,[XXXXX])]
 5: [(2@2,[XX])]
}
[0] lru_table_add(8)... 
extra: 2 
LRU_TABLE(4) 
{
LRU: [(3@3,[XXX]),(4@4,[XXXX]),(5@5,[XXXXX]),(8@8,[XXXXXXXX])]
 0: [(4@4,[XXXX]),(8@8,[XXXXXXXX])]
 4: [(3@3,[XXX]),(5@5,[XXXXX])]
}
[0] lru_table_add(9)... 
extra: 3 
LRU_TABLE(4) 
{
LRU: [(4@4,[XXXX]),(5@5,[XXXXX]),(8@8,[XXXXXXXX]),(9@9,[XXXXXXXXX])]
 0: [(4@4,[XXXX]),(8@8,[XXXXXXXX])]
 4: [(5@5,[XXXXX])]
 5: [(9@9,[XXXXXXXXX])]
}
search: 10
[0] lru_table_search(key=10)... 
LRU_TABLE(4) 
{
LRU: [(4@4,[XXXX]),(5@5,[XXXXX]),(8@8,[XXXXXXXX]),(9@9,[XXXXXXXXX])]
 0: [(4@4,[XXXX]),(8@8,[XXXXXXXX])]
 4: [(5@5,[XXXXX])]
 5: [(9@9,[XXXXXXXXX])]
}
[0] lru_table_add(11)... 
extra: 4 
LRU_TABLE(4) 
{
LRU: [(5@5,[XXXXX]),(8@8,[XXXXXXXX]),(9@9,[XXXXXXXXX]),(11@11,[XXXXXXXXXXX])]
 0: [(8@8,[XXXXXXXX])]
 4: [(5@5,[XXXXX]),(11@11,[XXXXXXXXXXX])]
 5: [(9@9,[XXXXXXXXX])]
}
search: 12
[0] lru_table_search(key=12)... 
LRU_TABLE(4) 
{
LRU: [(5@5,[XXXXX]),(8@8,[XXXXXXXX]),(9@9,[XXXXXXXXX]),(11@11,[XXXXXXXXXXX])]
 0: [(8@8,[XXXXXXXX])]
 4: [(5@5,[XXXXX]),(11@11,[XXXXXXXXXXX])]
 5: [(9@9,[XXXXXXXXX])]
}
[0] lru_table_add(13)... 
extra: 5 
LRU_TABLE(4) 
{
LRU: [(8@8,[XXXXXXXX]),(9@9,[XXXXXXXXX]),(11@11,[XXXXXXXXXXX]),(13@13,[XXXXXXXXXXXXX])]
 0: [(8@8,[XXXXXXXX])]
 3: [(13@13,[XXXXXXXXXXXXX])]
 4: [(11@11,[XXXXXXXXXXX])]
 5: [(9@9,[XXXXXXXXX])]
}
[0] lru_table_add(14)... 
extra: 8 
LRU_TABLE(4) 
{
LRU: [(9@9,[XXXXXXXXX]),(11@11,[XXXXXXXXXXX]),(13@13,[XXXXXXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
 3: [(13@13,[XXXXXXXXXXXXX])]
 4: [(11@11,[XXXXXXXXXXX])]
 5: [(9@9,[XXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
}
search: 15
[0] lru_table_search(key=15)... 
LRU_TABLE(4) 
{
LRU: [(9@9,[XXXXXXXXX]),(11@11,[XXXXXXXXXXX]),(13@13,[XXXXXXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
 3: [(13@13,[XXXXXXXXXXXXX])]
 4: [(11@11,[XXXXXXXXXXX])]
 5: [(9@9,[XXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
}
search: 16
[0] lru_table_search(key=16)... 
LRU_TABLE(4) 
{
LRU: [(9@9,[XXXXXXXXX]),(11@11,[XXXXXXXXXXX]),(13@13,[XXXXXXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
 3: [(13@13,[XXXXXXXXXXXXX])]
 4: [(11@11,[XXXXXXXXXXX])]
 5: [(9@9,[XXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
}
search: 17
[0] lru_table_search(key=17)... 
LRU_TABLE(4) 
{
LRU: [(9@9,[XXXXXXXXX]),(11@11,[XXXXXXXXXXX]),(13@13,[XXXXXXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
 3: [(13@13,[XXXXXXXXXXXXX])]
 4: [(11@11,[XXXXXXXXXXX])]
 5: [(9@9,[XXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
}
search: 18
[0] lru_table_search(key=18)... 
LRU_TABLE(4) 
{
LRU: [(9@9,[XXXXXXXXX]),(11@11,[XXXXXXXXXXX]),(13@13,[XXXXXXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
 3: [(13@13,[XXXXXXXXXXXXX])]
 4: [(11@11,[XXXXXXXXXXX])]
 5: [(9@9,[XXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
}
search: 19
[0] lru_table_search(key=19)... 
LRU_TABLE(4) 
{
LRU: [(9@9,[XXXXXXXXX]),(11@11,[XXXXXXXXXXX]),(13@13,[XXXXXXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
 3: [(13@13,[XXXXXXXXXXXXX])]
 4: [(11@11,[XXXXXXXXXXX])]
 5: [(9@9,[XXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX])]
}
[0] lru_table_add(20)... 
extra: 9 
LRU_TABLE(4) 
{
LRU: [(11@11,[XXXXXXXXXXX]),(13@13,[XXXXXXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX]),(20@20,[XXXXXXXXXXXXXXXXXXXX])]
 2: [(20@20,[XXXXXXXXXXXXXXXXXXXX])]
 3: [(13@13,[XXXXXXXXXXXXX])]
 4: [(11@11,[XXXXXXXXXXX])]
 5: [(14@14,[XXXXXXXXXXXXXX])]
}
search: 21
[0] lru_table_search(key=21)... 
LRU_TABLE(4) 
{
LRU: [(11@11,[XXXXXXXXXXX]),(13@13,[XXXXXXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX]),(20@20,[XXXXXXXXXXXXXXXXXXXX])]
 2: [(20@20,[XXXXXXXXXXXXXXXXXXXX])]
 3: [(13@13,[XXXXXXXXXXXXX])]
 4: [(11@11,[XXXXXXXXXXX])]
 5: [(14@14,[XXXXXXXXXXXXXX])]
}
[0] lru_table_add(22)... 
extra: 11 
LRU_TABLE(4) 
{
LRU: [(13@13,[XXXXXXXXXXXXX]),(14@14,[XXXXXXXXXXXXXX]),(20@20,[XXXXXXXXXXXXXXXXXXXX]),(22@22,[XXXXXXXXXXXXXXXXXXXXXX])]
 2: [(20@20,[XXXXXXXXXXXXXXXXXXXX])]
 3: [(13@13,[XXXXXXXXXXXXX])]
 4: [(22@22,[XXXXXXXXXXXXXXXXXXXXXX])]
 5: [(14@14,[XXXXXXXXXXXXXX])]
}
[0] lru_table_add(23)... 
extra: 13 
LRU_TABLE(4) 
{
LRU: [(14@14,[XXXXXXXXXXXXXX]),(20@20,[XXXXXXXXXXXXXXXXXXXX]),(22@22,[XXXXXXXXXXXXXXXXXXXXXX]),(23@23,[XXXXXXXXXXXXXXXXXXXXXXX])]
 0: [(23@23,[XXXXXXXXXXXXXXXXXXXXXXX])]
 2: [(20@20,[XXXXXXXXXXXXXXXXXXXX])]
 4: [(22@22,[XXXXXXXXXXXXXXXXXXXXXX])]
 5: [(14@14,[XXXXXXXXXXXXXX])]
}
search: 24
[0] lru_table_search(key=24)... 
LRU_TABLE(4) 
{
LRU: [(14@14,[XXXXXXXXXXXXXX]),(20@20,[XXXXXXXXXXXXXXXXXXXX]),(22@22,[XXXXXXXXXXXXXXXXXXXXXX]),(23@23,[XXXXXXXXXXXXXXXXXXXXXXX])]
 0: [(23@23,[XXXXXXXXXXXXXXXXXXXXXXX])]
 2: [(20@20,[XXXXXXXXXXXXXXXXXXXX])]
 4: [(22@22,[XXXXXXXXXXXXXXXXXXXXXX])]
 5: [(14@14,[XXXXXXXXXXXXXX])]
}
search: 25
[0] lru_table_search(key=25)... 
LRU_TABLE(4) 
{
LRU: [(14@14,[XXXXXXXXXXXXXX]),(20@20,[XXXXXXXXXXXXXXXXXXXX]),(22@22,[XXXXXXXXXXXXXXXXXXXXXX]),(23@23,[XXXXXXXXXXXXXXXXXXXXXXX])]
 0: [(23@23,[XXXXXXXXXXXXXXXXXXXXXXX])]
 2: [(20@20,[XXXXXXXXXXXXXXXXXXXX])]
 4: [(22@22,[XXXXXXXXXXXXXXXXXXXXXX])]
 5: [(14@14,[XXXXXXXXXXXXXX])]
}
search: 26
[0] lru_table_search(key=26)... 
LRU_TABLE(4) 
{
LRU: [(14@14,[XXXXXXXXXXXXXX]),(20@20,[XXXXXXXXXXXXXXXXXXXX]),(22@22,[XXXXXXXXXXXXXXXXXXXXXX]),(23@23,[XXXXXXXXXXXXXXXXXXXXXXX])]
 0: [(23@23,[XXXXXXXXXXXXXXXXXXXXXXX])]
 2: [(20@20,[XXXXXXXXXXXXXXXXXXXX])]
 4: [(22@22,[XXXXXXXXXXXXXXXXXXXXXX])]
 5: [(14@14,[XXXXXXXXXXXXXX])]
}
[0] lru_table_add(27)... 
extra: 14 
LRU_TABLE(4) 
{
LRU: [(20@20,[XXXXXXXXXXXXXXXXXXXX]),(22@22,[XXXXXXXXXXXXXXXXXXXXXX]),(23@23,[XXXXXXXXXXXXXXXXXXXXXXX]),(27@27,[XXXXXXXXXXXXXXXXXXXXXXXXXXX])]
 0: [(23@23,[XXXXXXXXXXXXXXXXXXXXXXX])]
 2: [(20@20,[XXXXXXXXXXXXXXXXXXXX])]
 4: [(22@22,[XXXXXXXXXXXXXXXXXXXXXX]),(27@27,[XXXXXXXXXXXXXXXXXXXXXXXXXXX])]
}
[0] lru_table_add(28)... 
extra: 20 
LRU_TABLE(4) 
{
LRU: [(22@22,[XXXXXXXXXXXXXXXXXXXXXX]),(23@23,[XXXXXXXXXXXXXXXXXXXXXXX]),(27@27,[XXXXXXXXXXXXXXXXXXXXXXXXXXX]),(28@28,[XXXXXXXXXXXXXXXXXXXXXXXXXXXX])]
 0: [(23@23,[XXXXXXXXXXXXXXXXXXXXXXX])]
 2: [(28@28,[XXXXXXXXXXXXXXXXXXXXXXXXXXXX])]
 4: [(22@22,[XXXXXXXXXXXXXXXXXXXXXX]),(27@27,[XXXXXXXXXXXXXXXXXXXXXXXXXXX])]
}
[0] lru_table_add(29)... 
extra: 22 
draining...
polled: 23
[XXXXXXXXXXXXXXXXXXXXXXX][0] keyvalue_destroy()... 
polled: 27
[XXXXXXXXXXXXXXXXXXXXXXXXXXX][0] keyvalue_destroy()... 
polled: 28
[XXXXXXXXXXXXXXXXXXXXXXXXXXXX][0] keyvalue_destroy()... 
polled: 29
[XXXXXXXXXXXXXXXXXXXXXXXXXXXXX][0] keyvalue_destroy()... 
[0] lru_table_free()... 

Normal exit.
